refactor: Remove SIPTransport and replace it with UnityTransport

NoelStephensUnity opened the pull request:
This removes the SIPTransport from our integration tests and replaces it with UnityTransport.
This includes several integration test refactors, and some minor fixes to bugs that were not detected under SIPTransport.

[MTT-3016](https://jira.unity3d.com/browse/MTT-3016)

## Changelog
- Changed: NetcodeIntegrationTestHelpers to use UnityTransport.
- Removed SIPTransport from integration tests.

## Testing and Documentation
- Includes adjustments to integration tests.

0xFA11 committed:
delete `SIPTransport` &  `SIPTransportTests`

0xFA11 committed:
fix compile errors

NoelStephensUnity committed:
fix

This fixes the timing issues with the MessageOrdering.SpawnRpcDespawn integration test.
This also includes a minor update to NetcodeIntegrationTestHelpers (destroy as opposed to destroy immediate).

NoelStephensUnity committed:
fix

This fixes the issue with MultiClientConnectionApproval.ClientConnectedCallbackTest failing when using UTP as opposed to SIP.

NoelStephensUnity committed:
fix

Fixed issue with ConnectionApprovalHandler's prefab getting destroyed when a client was disconnected.
Applied same fix to the NetcodeIntegrationTest when creating a prefab.

Also fixing issue with NetworkManager spewing spam messages if the NetworkConfig.ConnectionApproval is set but there is no ConnectionApprovalCallback assigned.  Adjusted it to make sure it is either the Server or Host that this applies to, but not a client.

NoelStephensUnity committed:
fix

This fixes the ConnectionApprovalMismatchTest issue that caused it to fail with UTP.
Switching over to client notifications (i.e. the server detects the mismatch and disconnects the clients, we want to make sure the clients received the disconnection notification).

NoelStephensUnity committed:
fix

Global fix for teardown and NetworkObject destroying....make sure the server is the NetworkManagerOwner before destroying it.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/kill-sip-use-utp

NoelStephensUnity committed:
fix

Some NetcodeIntegrationTest adjustments to handle the switch over to UTP.

NoelStephensUnity committed:
fix

Fixes issues with NetworkObjectParentingTests (increased wait time between changes to 1 full Network Tick).
Also the recent NetcodeIntegrationTest updates helped fix some of the issues with NetworkObjectParentingTests.

NoelStephensUnity committed:
fix

Fixes the issue with NetworkSpawnManagerTests.TestConnectAndDisconnect.

NoelStephensUnity committed:
Fix - NetworkSceneManager

Switching to UnityTransport exposed a potential bug where the server disconnects a client while the client is still processing a SceneEventType.Synchronize message which by the time the client processes this it has already sent the SceneEventType.SynchronizeComplete message back to the server that *could* attempt to send a resynchronization message where the MessageSystem will crash because the m_SendQueues entry for the now disconnected client no longer exists.

NoelStephensUnity committed:
Fix - NetcodeIntegrationTest

Added some additional UnityTransport settings for payload size, send queue size, and reducing the connection attempts and time out to connect per attempt.

Added the ability to specify whether you want the server or clients to be instantiated first (see comments under NetcodeIntegrationTest.m_CreateServerFirst for further explanation).

NoelStephensUnity committed:
Fix - ClientOnlyConnectionTests

Removing UNet and just using UTP.
Reducing the timeout period so it doesn't take over 60 seconds to timeout.

NoelStephensUnity committed:
Fix - DisconnectTests

Fixes timing issues with this test that is causing it to not passing when using UTP.

NoelStephensUnity committed:
fix - NetworkObjectOwnershipTests

Fixed timing issues exposed when using UTP.

NoelStephensUnity committed:
fix - NetworkSceneManagerEventDataPoolTest

This fixes an issue when I originally switched it to be the server as the default NetworkManager to be instantiated when using NetcodeIntegrationTestHelper.Create.  This sets the, recently updated NetcodeIntegrationTest  m_CreateServerFirst, clients to instantiate before the server (needed for in-scene objects)

NoelStephensUnity committed:
fix - NetworkSceneManagerSeneVerification

This test requires the clients to be instantiated before the server.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
style

NoelStephensUnity committed:
update MTT-3016

MTT-3016
Changelog entries.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Update CHANGELOG.md

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -17,11 +17,11 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ### Changed
-
+- Changed NetcodeIntegrationTestHelpers to use UnityTransport. (#1870)
 - Updated `UnityTransport` dependency on `com.unity.transport` to 1.0.0 (#1849)
 
 ### Removed
-
+- Removed SIPTransport from integration tests. (#1870)
```
we probably do not need to have these changelog entries.

NoelStephensUnity replied:
NetcodeIntegrationTest is public facing... as well as other integration test related code is public facing too...so it might be just "useful to know"...? (at least that was my logic)

0xFA11 replied:
was `SIPTransport` ever public? if so, it does make sense to announce its removal.

0xFA11 replied:
oh wow, apparently it **is** public:
https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/b5f4195c8ce26294ff6f29d9870e3d546628ff92/com.unity.netcode.gameobjects/TestHelpers/Runtime/Transport/SIPTransport.cs#L13

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1746,7 +1746,9 @@ private void HandleServerSceneEvent(uint sceneEventId, ulong clientId)
                         m_NetworkManager.InvokeOnClientConnectedCallback(clientId);
 
-                        if (sceneEventData.ClientNeedsReSynchronization() && !DisableReSynchronization)
+                        // Check to see if the client needs to resynchronize and before sending the message make sure the client is still connected to avoid
+                        // a potential crash within the MessageSystem (i.e. sending to a client that no longer exists)
+                        if (sceneEventData.ClientNeedsReSynchronization() && !DisableReSynchronization && m_NetworkManager.ConnectedClients.ContainsKey(clientId))
```
@ShadauxCat ?

0xFA11 commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -79,6 +79,15 @@ public static void DeregisterNetworkObject(ulong localClientId, ulong networkObj
         protected const uint k_DefaultTickRate = 30;
         protected abstract int NumberOfClients { get; }
 
+        /// <summary>
+        /// Set this to false to create the clients first.
+        /// Note: If you are using scene placed NetworkObjects or doing any form of scene testing and
+        /// get prefab hash id "soft synchronization" errors, then set this to false and run your test
+        /// again.  This is a work-around until we can resolve some issues with NetworkManagerOwner and
+        /// NetworkManager.Singleton.
+        /// </summary>
```
do you mind elaborating a little more? what's going on here?

NoelStephensUnity replied:
Its a little complicated, but it has to do with our integration tests running in the same thread, how we populate newly loaded (not yet "spawned") in-scene placed NetworkObjects, and NetworkObject.NetworkManager referencing NetworkManager.Singleton if NetworkObject.NetworkManagerOwner is null.

- Same Thread: In several places we check to make sure the NetworkObject.NetworkManager is the "right one" for the NetworkManager context (i.e. NetworkObject.NetworkManager == NetworkManager)
- Populating In-Scene Objects:  This checks to make sure the NetworkSceneManager.m_NetworkManager is the same as the NetworkObject.NetworkManager before adding it to the ScenePlacedObjects list.
- Default NetworkManager:  This happens to be NetworkManager.Singleton that is set upon the first NetworkManager being instantiated.

So, under some conditions you might want NetworkManager.Singleton to be the server and under other conditions you might want it to be a client.  If you are loading a scene with in-scene placed NetworkObjects, then you will want the NetworkManager.Singleton to be the first client instantiated in order to properly synchronize/find the in-scene placed NetworkObjects (which are loaded by the server and not loaded by the client since they share the same scene space).

If you want more details feel free to DM me in slack... really we need to handle in-scene placed NetworkObjects better... like that "pre-rfc" document outlines.

0xFA11 commented on testproject/Assets/Tests/Runtime/MessageOrdering.cs:
```diff
@@ -103,60 +107,78 @@ public IEnumerator SpawnChangeOwnership()
+            var allHandlersDestroyed = false;
+            var waitForTick = new WaitForSeconds(1.0f / m_ServerNetworkManager.NetworkConfig.TickRate);
+
+            while (!(allHandlersSpawned && clientCountReached && allHandlersDestroyed))
```
I'd add @ShadauxCat as a reviewer to review changes around these tests.

0xFA11 replied:
(just in case)

NoelStephensUnity committed:
fix - tools

This resolves the issues with tools integration tests and switching over to UnityTransport.

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -17,11 +17,11 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Removed
-
+- Removed SIPTransport from integration tests. (#1870)
```
minor:

```suggestion

- Removed `SIPTransport` (#1870)
```

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -17,11 +17,11 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Changed
-
+- Changed NetcodeIntegrationTestHelpers to use UnityTransport. (#1870)
```
minor:

```suggestion

- Changed `NetcodeIntegrationTestHelpers` to use `UnityTransport` (#1870)
```

0xFA11 reviewed (approved):
lgtm 🚀

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/CHANGELOG.md

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/CHANGELOG.md

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
fix

The following fixes some edge-case timing related issues that are being exposed in Yamato where it might take longer than 1 network tick depending upon the virtual machine's current load.  Primarily, the majority of the fixes are handled by just waiting for the condition or timing out.

NoelStephensUnity committed:
fix - console - ps4

the time out was never yielding and causing issues on PS4/

Rosme reviewed (approved):
I'm fine with the changes in the tests integration stuff as well as whatever may touch the tools.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/kill-sip-use-utp

NoelStephensUnity committed:
revert

Removing some code that didn't need to make it into this PR.
